client/protoclient.go

```
@@ -0,0 +1,491 @@
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func new(docslogLevel logrus.Level, requestTimeout ...time.Duration) *ProtoClient {
+	return &ProtoClient{
+		Client:           *New(docslogLevel, requestTimeout...),
+		descriptorsNames: make(map[string]bool),
+		info: ProtoBufferInfo{
+			Commands: make(map[string]*Command),
+		},
+		docsRoute:        "",
+		descriptorsRoute: "",
+		IncomingMsgChan:  make(chan *message.Message, 3),
+		closeChan:        make(chan bool),
+	}
+}
+
+func NewProto(docsRoute string, docslogLevel logrus.Level, requestTimeout ...time.Duration) *ProtoClient {
+	newclient := new(docslogLevel, requestTimeout...)
+	newclient.docsRoute = docsRoute
+	return newclient
+}
+
+func NewWithDescriptor(descriptorsRoute string, docsRoute string, docslogLevel logrus.Level, requestTimeout ...time.Duration) *ProtoClient {
+	newclient := new(docslogLevel, requestTimeout...)
+	newclient.docsRoute = docsRoute
+	newclient.descriptorsRoute = descriptorsRoute
+	return newclient
+}
+
+// Load commands information form the server. Names is a list of protos names.
+func (pc *ProtoClient) LoadServoInfo(addr string) error {
+	pc.ready = false
+
+	if err := pc.ConnectToTLS(addr, true); err != nil {
+		if err.Error() == "EOF" {
+			if err := pc.ConnectTo(addr); err != nil {
+				return err
+			}
+		} else {
+			return err
+		}
+	}
+
+	// request doc info
+	_, err := pc.SendRequest(pc.docsRoute, make([]byte, 0))
+	if err != nil {
+		return err
+	}
+	response := <-pc.Client.IncomingMsgChan
+
+	docs := &protos.Doc{}
+	if err := proto.Unmarshal(response.Data, docs); err != nil {
+		return err
+	}
+
+	if err := pc.getDescriptors(docs.Doc); err != nil {
+		return err
+	}
+
+	pc.Disconnect()
+	pc.ready = true
+
+	return nil
+}
+
+// Disconnect the client
+func (pc *ProtoClient) Disconnect() {
+	pc.Client.Disconnect()
+	if pc.ready {
+		pc.closeChan <- true
+	}
+}
+
+func (pc *ProtoClient) waitForData() {
+	if !pc.ready {
```

Wouldn't it be better to initialize the client if it isn't yet instead of silently returning?